### ISSUE TITLE
Idea for new 'setCountAndListRef' Create.hs function

### ISSUE THREAD

**Issue by Rotaerk** [achirkin/vulkan#12]
I've noticed a lot of structures in which you have to set a count field and a pointer field, and it often ends up looking like this:

```hs
set @"enabledExtensionCount" (lengthNum extensions) &*
setListRef @"ppEnabledExtensionNames" extensions
```

where:

```hs
lengthNum :: (Foldable t, Num n) => t a -> n
lengthNum = fromIntegral . length
{-# INLINE lengthNum #-}
```

I'm now imagining a helper for this pattern, something that looks like this:

```hs
setCountAndListRef @"enabledExtensionCount" @"ppEnabledExtensionNames" extensions
```

Best I can come up with to implement it is this:

```hs
setCountAndListRef ::
  forall countFname arrayFname x c s.
  (
    CanWriteField countFname x,
    CanWriteField arrayFname x,
    FieldType countFname x ~ c,
    FieldType arrayFname x ~ Ptr s,
    Storable s,
    Num c
  ) =>
  [s] -> CreateVkStruct x '[countFname, arrayFname] ()
setCountAndListRef ss =
  set @countFname (lengthNum ss) &*
  setListRef @arrayFname ss
```

But it doesn't type-check, and I'm unsure how to resolve it.  It complains that it can't deduce `Graphics.Vulkan.Marshal.Create.Union x '[countFname] '[arrayFname] ~ '[countFname, arrayFname]`.

Edit: Maybe a better name would be `setListCountAndRef`.

**Comment by achirkin** [comment]
Ah, I see.
The problem is that GHC cannot rule out the possibility that `countFname ~ arrayFname` for  the [`Union`](https://github.com/achirkin/vulkan/blob/abe9b65df6b776a66ca0e1a14c094b6187adf1bc/vulkan-api/src/Graphics/Vulkan/Marshal/Create.hs#L246) type family. So it cannot figure out that union of the two elements is indeed a list of two elements. Unfortunately, there is no way in current GHC to provide constraint that the two types are NOT the same. Maybe, I need to rethink the whole `CreateVk` class in the end :(

At this moment, the workaround is to add `Graphics.Vulkan.Marshal.Create.Union x '[countFname] '[arrayFname] ~ '[countFname, arrayFname]` to the constraints of your function. Looks clumsy, but should work.

**Comment by Rotaerk** [comment]
Forgot I created this issue.  It's been resolved with #22.
